cmd/zc_traverser_blob.go

```diff
@@ -236,13 +239,83 @@ func (t *blobTraverser) Traverse(preprocessor objectMorpher, processor objectPro
+	// CFDMode is to enumerate Target and compare. So we return true in that case.
+	// This will lead to enumeration of target.
+	if t.cfdMode == common.CFDModeFlags.TargetCompare() {
+		return true
+	}
+
+	if t.cfdMode == common.CFDModeFlags.CtimeMtime() {
+		if so.lastChangeTime.After(t.lastSyncTime) {
+			return true
+		}
+		return false
+	} else if t.cfdMode == common.CFDModeFlags.Ctime() {
+		if so.lastChangeTime.After(t.lastSyncTime) {
+			return true
+		} else if t.metaDataOnlySync && t.indexerMap.filesChangedInDirectory(so.relativePath, t.lastSyncTime) {
+			// Listing of blob gives properties and each list api gives 5K blobs in single rest call. So if we know any file changed on source,
+			// We should list the target. It's optimization instead of getting properties for changed files one by one get properties of all blob.
+			// As in normal scenario, there will be less than 2K files in a folder. It make sense to call list api, if any file changed on source.
+			return true
+		} else {
+			return false
+		}
+	} else {
+		// Lets try to get the properties for directory.
+		blobURL := containerURL.NewBlobURL(strings.TrimSuffix(so.relativePath, common.AZCOPY_PATH_SEPARATOR_STRING))
+		resp, err := blobURL.GetProperties(context.TODO(), azblob.BlobAccessConditions{}, azblob.ClientProvidedKeyOptions{})
+		if err != nil {
+			return true
+		}
+		// Note: ContentLength gives the size of the blob in bytes. For a page blob, this header returns the value of the x-ms-blob-content-length header that is stored with the blob.
+		if so.lastModifiedTime.After(resp.LastModified()) || so.size != resp.ContentLength() {
+			return true
+		} else {
+			return false
+		}
+	}
+}
+
 func (t *blobTraverser) parallelList(containerURL azblob.ContainerURL, containerName string, searchPrefix string,
 	extraSearchPrefix string, preprocessor objectMorpher, processor objectProcessor, filters []ObjectFilter) error {
 	// Define how to enumerate its contents
 	// This func must be thread safe/goroutine safe
 	enumerateOneDir := func(dir parallel.Directory, enqueueDir func(parallel.Directory), enqueueOutput func(parallel.DirectoryEntry, error)) error {
+		if _, ok := dir.(string); !ok {
+			panic("Directory entry is not string")
+		}
+
 		currentDirPath := dir.(string)
 
+		// This code for sync operation and when its target traverser.
+		if t.isSync && !t.isSource {
+			so := t.indexerMap.getStoredObject(currentDirPath)
+			currentDirPath = so.relativePath
+
+			// source traverser (local traverser) truncate all the forward slash in storedObject.
+			// Without forward slash list blob not list blobs under this directory.
+			if currentDirPath != "" {
+				currentDirPath = currentDirPath + "/"
+			}
+
+			// Check the directory if we need enumeration at target side or not.
+			if !t.HasDirectoryChangedSinceLastSync(so, containerURL) {
+				goto FinalizeDirectory
+			}
```

Comment:
we should use the same FinalizeAll boolean flag as in the doc and convey it using the storedObject enqueued at L429.
Let's discuss this.